Makefile

```diff
@@ -47,43 +27,80 @@ GO_BUILD_LDFLAGS          := $(shell $(GO_BUILD_LDFLAGS_CMD) LDFLAG)
+THRIFTRW_BIN = $(BUILD)/bin/thriftrw
+COPYRIGHT_BIN = $(BUILD)/bin/copyright
+MOCKGEN_BIN = $(BUILD)/bin/mockgen
+ENUMER_BIN = $(BUILD)/bin/enumer
+GOIMPORTS_BIN = $(BUILD)/bin/goimports
+GOLINT_BIN = $(BUILD)/bin/golint
+
+# downloads and builds a go-gettable tool, versioned by go.mod, and installs
+# it into the build folder, named the same as the last portion of the URL.
+define get_tool
+@echo "building $(1) into $(BUILD)/bin/$(notdir $(1)) ..."
+@go build -mod=readonly -o $(BUILD)/bin/$(notdir $(1)) $(1)
 endef
 
-$(foreach tsrc,$(THRIFTRW_SRCS),$(eval $(call \
-	thriftrwrule,$(basename $(notdir \
-	$(shell echo $(tsrc) | tr A-Z a-z))),$(tsrc))))
-
-# Automatically gather all srcs
-ALL_SRC := $(shell find . -name "*.go" | grep -v -e Godeps -e vendor \
-	-e ".*/\..*" \
-	-e ".*/_.*" \
-	-e ".*/mocks.*")
-
-# filter out the src files for tools
-TOOLS_SRC := $(shell find ./tools -name "*.go")
-TOOLS_SRC += $(TOOLS_CMD_ROOT)
-
+$(THRIFTRW_BIN): go.mod
+	$(call get_tool,go.uber.org/thriftrw)
+	$(call get_tool,go.uber.org/yarpc/encoding/thrift/thriftrw-plugin-yarpc)
+
+$(COPYRIGHT_BIN): cmd/tools/copyright/licensegen.go
+	go build -o $(COPYRIGHT_BIN) ./cmd/tools/copyright/licensegen.go
+
+$(MOCKGEN_BIN): go.mod
+	$(call get_tool,github.com/golang/mock/mockgen)
+
+$(ENUMER_BIN): go.mod
+	$(call get_tool,github.com/dmarkham/enumer)
+
+$(GOIMPORTS_BIN): go.mod
+	$(call get_tool,golang.org/x/tools/cmd/goimports)
+
+$(GOLINT_BIN): go.mod
+	$(call get_tool,golang.org/x/lint/golint)
+
+
+THRIFT_GENDIR=.gen/go
+THRIFT_SRCS := $(shell find idls -name '*.thrift')
+# concrete targets to build / the "sentinel" go files that need to be produced per thrift file.
+# idls/thrift/thing.thrift -> thing.thrift -> thing -> .gen/go/thing/thing.go
+THRIFT_GEN_SRC := $(foreach tsrc,$(basename $(subst idls/thrift/,,$(THRIFT_SRCS))),$(THRIFT_GENDIR)/$(tsrc)/$(tsrc).go)
+
+# how to generate each thrift file.
+# note that each generated file depends on ALL thrift files - this is necessary because they can import each other.
+$(THRIFT_GEN_SRC): $(THRIFT_SRCS) $(THRIFTRW_BIN)
+	@# .gen/go/thing/thing.go -> thing.go -> "thing " -> thing -> idls/thrift/thing.thrift
+	$(THRIFTRW_BIN) --plugin=yarpc --pkg-prefix=$(PROJECT_ROOT)/$(THRIFT_GENDIR) --out=$(THRIFT_GENDIR) --no-recurse idls/thrift/$(strip $(basename $(notdir $@))).thrift
+
+# Automatically gather all srcs.
+# Works by ignoring everything in the parens (and does not descend into matching folders) due to `-prune`,
+# and everything else goes to the other side of the `-o` branch, which is `-print`ed.
+# This is dramatically faster than a `find . | grep -v vendor` pipeline.
+ALL_SRC := $(shell \
+	find . \
+	\( \
+		-path './vendor/*' \
+		-o -path './.*' \
+		-o -path '*/mocks*' \
+	\) \
+	-prune \
+	-o -name '*.go' -print \
+)
+
+# make sure a sentinel thrift-generated file is in ALL_SRC, so thrift is (re)generated if necessary
+ALL_SRC += $(THRIFT_GEN_SRC)
+FMT_SRC := $(filter-out .gen/%, $(ALL_SRC))
```

Comment:
Does this mean we will not run goimport for generated files?

Reply:
when I was running all this, that seemed to be the case currently.  if I ran goimports on those files, it moved imports around a bit.
possibly that's just a different version of goimports though, I'm not quite sure.  I'll look into it a bit more and find out exactly what's happening.

generally speaking though I don't feel strongly either way.  not-running it is potentially simpler (no need to consider it for task sequencing), but running it does avoid accidental IDE tweaks.  would you / we prefer to goimports it all?

Reply:
Since we already do modifications to generated files by adding license header, I think it also makes sense to modify it further with goimports. So yeah, I would vote to run goimport on generated files as well.